specs/core/1_shard-data-chains.md

```diff
@@ -222,3 +222,243 @@ The `shard_chain_commitment` is only valid if it equals `compute_commitment(head
+        domain=get_domain(state, get_current_epoch(state), DOMAIN_CUSTODY_INTERACTIVE)
+    )
+    # Check that the responder actually participated in the attestation
+    assert responder_index in attestation.validator_indices
+    # Can only be challenged by one challenger at a time
+    assert responder.interactive_custody_challenge_data.challenger_index == VALIDATOR_NULL
+    # Can only challenge one responder at a time
+    assert challenger.now_challenging == VALIDATOR_NULL
+    # Can't challenge if you've been penalized
+    assert challenger.penalized_epoch == FAR_FUTURE_EPOCH
+    # Make sure the revealed subkey is valid
+    assert verify_custody_subkey_reveal(
+        pubkey=state.validator_registry[responder_index].pubkey,
+        subkey=responder_subkey,
+        mask=ZERO_HASH,
+        mask_pubkey=b'',
+        period=slot_to_custody_period(attestation.data.slot)
+    )
+    # Set the challenge object
+    responder.interactive_custody_challenge_data = InteractiveCustodyChallengeData(
+        challenger=initiation.challenger_index,
+        data_root=attestation.custody_commitment,
+        custody_bit=get_bitfield_bit(attestation.custody_bitfield, attestation.validator_indices.index(responder_index)),
+        responder_subkey=responder_subkey,
+        current_custody_tree_node=ZERO_HASH,
+        depth=0,
+        offset=0,
+        max_depth=get_merkle_depth(initiation.attestation),
+        deadline=get_current_epoch(state) + CHALLENGE_RESPONSE_DEADLINE
+    )
+    # Responder can't withdraw yet!
+    state.validator_registry[responder_index].withdrawable_epoch = FAR_FUTURE_EPOCH
+    # Challenger can't challenge anyone else
+    challenger.now_challenging = responder_index
+```
+
+We define an `InteractiveCustodyChallengeResponse` as follows:
+
+```python
+{
+    'responder_index': 'uint64',
+    'hashes': ['bytes32'],
+    'signature': 'bytes96',
+}
+```
+
+A response provides 32 hashes that are under current known proof of custody tree node. Note that at the beginning the tree node is just one bit of the custody root, so we ask the responder to sign to commit to the top 5 levels of the tree and therefore the root hash; at all other stages in the game responses are self-verifying.
+
+Here's the function for verifying and processing a response:
+
+```python
+def process_response(response: InteractiveCustodyChallengeResponse,
+                     state: State):
+    responder = state.validator_registry[response.responder_index]
+    challenge_data = responder.interactive_custody_challenge_data
+    # Check that the right number of hashes was provided
+    expected_depth = min(challenge_data.max_depth - challenge_data.depth, MAX_POC_RESPONSE_DEPTH)
+    assert 2**expected_depth == len(response.hashes)
+    # Must make some progress!
+    assert expected_depth > 0
+    # Check the hashes match the previously provided root
+    root = merkle_root(hashes)
+    if challenge_data.depth == 0:
```

Comment:
In this case, we should set the `challenge_data.current_custody_tree_node = root`, correct?

Reply:
Because the responder has committed to it and it is needed when calling `process_continuation`